### ISSUE TITLE
About the comments about coordinate systems in tv_wrapper.py

### ISSUE THREAD

**Issue by silencht** [unitreerobotics/avp_teleoperate#25]
This issue is created to address the branching problem raised in issue #19.
The problem is described as follows:

Hi I am reading the comments about coordinate systems in [tv_wrapper](https://github.com/unitreerobotics/avp_teleoperate/blob/g1/teleop/open_television/tv_wrapper.py). I am a little bit confused by the wrist coordinates as stated in the comments: 
>  
     (Left Wrist) XR/AppleVisionPro Convention:
        - the x-axis pointing from wrist toward middle.
        - the y-axis pointing from index toward pinky.
        - the z-axis pointing from palm toward back of the hand.

     (Right Wrist) XR/AppleVisionPro Convention:
        - the x-axis pointing from wrist toward middle.
        - the y-axis pointing from pinky toward index.
        - the z-axis pointing from palm toward back of the hand.
        
     p.s. From website: https://registry.khronos.org/OpenXR/specs/1.1/man/html/openxr.html.
         You can find **(Left/Right Wrist) XR/AppleVisionPro Convention** related information like this below:
           "The wrist joint is located at the pivot point of the wrist, which is location invariant when twisting the hand without moving the forearm. 
            The backward (+Z) direction is parallel to the line from wrist joint to middle finger metacarpal joint, and points away from the finger tips. 
            The up (+Y) direction points out towards back of the hand and perpendicular to the skin at wrist. 
            The X direction is perpendicular to the Y and Z directions and follows the right hand rule."
         Note: The above context is of course under **(basis) OpenXR Convention**.

The coordinates of both hands do not align with the website statement. Can you explain what is the difference?

**Comment by silencht** [comment]
@BBBBBBob

For this issue, a reference image is provided for your convenience. If you have any further questions, please feel free to add them under this issue. Thank you!


Taking the left hand as an example:

![left_hand](https://github.com/user-attachments/assets/e0adbb84-0a17-43e9-adbd-f3696747911e)

- Red
> under (basis) Robot Convention, wrist / hand's initial pose convention:
(Left Wrist) XR/AppleVisionPro Convention:
        - the x-axis pointing from wrist toward middle.
        - the y-axis pointing from index toward pinky.
        - the z-axis pointing from palm toward back of the hand.
- Blue
> under (basis) OpenXR Convention 
The backward (+Z) direction is parallel to the line from wrist joint to middle finger metacarpal joint, and points away from the finger tips. 
        The up (+Y) direction points out towards back of the hand and perpendicular to the skin at wrist. 
        The X direction is perpendicular to the Y and Z directions and follows the right hand rule.

**Comment by BBBBBBob** [comment]
@silencht Thank you for the answer! I am still confused about the Unitree convention of hand. Assuming that the coordinate of robot convention is fixed, is the initial position of Unitree hand different from the VR position or is the coordinate system different? Because the coordinate of our hands varies from the example, I want to be clear how `T_to_unitree_hand` is calculated.

**Comment by silencht** [comment]
@BBBBBBob
In short, the role of `tv_wrapper` is to transform the OpenXR convention into the Robot convention commonly used in robotics.

As for some of the other content, using the left hand as an example:  
The purpose of `T_to_unitree_left_wrist` is simply to rotate the initial coordinate system by 90°. This is because, whether in the OpenXR or Robot convention, the initial orientation (represented by the identity matrix) assumes that the palm faces downward, whereas the palm of the Unitree left hand faces to the right.

As for the `T_to_unitree_hand` you mentioned, it is related to the usage of the [dex_retargeting](https://github.com/dexsuite/dex-retargeting) library. 
I recommend that you refer to https://github.com/dexsuite/dex-retargeting/issues/13 for further details.

**Comment by BBBBBBob** [comment]
@silencht Thank you for pointing out! `T_to_unitree_hand` should be a transformation matrix mapping the OpenXR convention to the hand convention. Is it correct?

**Comment by BBBBBBob** [comment]
Under the robot convention, does the unitree convention look like the following pose:

![left_hand](https://github.com/user-attachments/assets/2a67205b-da16-4d73-8da8-9adf4470218d)

so that
> 
(Left Hand URDF) Unitree Convention:   
        - The x-axis pointing from palm toward back of the hand. 
        - The y-axis pointing from middle toward wrist.
        - The z-axis pointing from pinky toward index.
      
is satisfied.

**Comment by silencht** [comment]
Yes. What you drew is correct.
However, essentially, I made some modifications to the URDF of the UFACTORY dexterous hand to make its coordinate system consistent with the previous Inspire dexterous hand URDF.
You can see that I applied a rotation at this position in the URDF file(The same applies to the right-hand URDF):
https://github.com/unitreerobotics/avp_teleoperate/blob/0b598c115306c6ad6a1ce588799ce32dc628999f/assets/unitree_hand/unitree_dex3_left.urdf#L257-L261

Perhaps I can use a diagram to explain this:
![hand](https://github.com/user-attachments/assets/8de0832e-458b-427f-8660-cdb374d4a0ea)

To be honest, explaining these coordinate transformations is indeed quite troublesome. I hope that in the future, I can explain all these changes in the form of documentation. However, this will take time, and the timeline is uncertain.

If you have any further questions, feel free to ask here.

**Comment by BBBBBBob** [comment]
Indeed, I have already solved the transformation based on our hand's urdf. Take your image for example:

![image](https://github.com/user-attachments/assets/bc4d78b7-abd8-4bcc-86dc-5ad0cf614790)

The green coordinate is the unitree hand convention as described above. I first adjusted the hand configuration as shown in the image and then calculated the rotation matrix between the red and green coordinates, which also works for our urdf coordinate. I hope this way is easier to understand.

**Comment by Aloys-y** [comment]
@BBBBBBob hi! I am also confused by the wrist coordinates as stated in the comments. Could I get in touch with you and ask for your advice? This is my email address 2625807792@qq.com

**Comment by BBBBBBob** [comment]
> [@BBBBBBob](https://github.com/BBBBBBob) hi! I am also confused by the wrist coordinates as stated in the comments. Could I get in touch with you and ask for your advice? This is my email address [2625807792@qq.com](mailto:2625807792@qq.com)

Hi, you can leave your questions here

**Comment by Aloys-y** [comment]
The coordinate system convention for OpenXR is assumed to be:
  Left wrist: +x points toward thumb, +y points from palm to back, +z points from middle finger to wrist
  Right wrist: +x points toward pinky, +y points from palm to back, +z points from middle finger to wrist
Note: I am not entirely sure this OpenXR coordinate system assumption is correct, which might cause misalignment with SMPL-H.

I am trying to integrate the captured hand rotations into an SMPL-H model. The primary goal is finger motion reconstruction, i.e., the finger joints should bend/extend correctly relative to the wrist.

Problem / Observed Behavior:
When computing local rotations relative to the wrist (R_local = R_parent^T @ R_child), the fingers do not bend correctly.
Instead, the fingers seem to twist in place rather than bending toward the palm.
I suspect this is because the initial wrist coordinate system in OpenXR differs from the SMPL-H rest pose coordinate system, and/or my assumption about OpenXR axes might be incorrect.
Even after applying a simple wrist alignment transformation, the finger directions remain incorrect.

Question:
How can I map the VisionPro/OpenXR global hand rotations to SMPL-H hand rotations so that:
The wrist rest pose aligns with SMPL-H.
Finger joint rotations are correctly computed relative to the wrist.

[xr_to_smplh.py](https://github.com/user-attachments/files/22405498/xr_to_smplh.py)

Thank you in advance for any guidance or suggestions! 🙏
